Fix V4 signed policy docs for MSVC

When we try to enable the code to convert character sets to UTF-8 we get:

https://source.cloud.google.com/results/invocations/431eb776-4c0f-43d1-8b47-b4d4f404541f

```
storage_client.lib(policy_document_request.obj) : error LNK2001: unresolved external symbol "__declspec(dllimport) public: static class std::locale::id std::codecvt<char32_t,char,struct _Mbstatet>::id" (__imp_?id@?$codecvt@_UDU_Mbstatet@@@std@@2V0locale@2@A)
  Hint on symbols that are defined and could potentially match:
    "__declspec(dllimport) public: static class std::locale::id std::codecvt<char,char,struct _Mbstatet>::id" (__imp_?id@?$codecvt@DDU_Mbstatet@@@std@@2V0locale@2@A)
    "__declspec(dllimport) public: static class std::locale::id std::time_get<char,class std::istreambuf_iterator<char,struct std::char_traits<char> > >::id" (__imp_?id@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@2V0locale@2@A)
bazel-out\x64_windows-fastbuild\bin\google\cloud\storage\examples\storage_notification_samples.exe : fatal error LNK1120: 1 unresolved externals
```

Sounds like we are using the wrong character type under MSVC?

--- a/google/cloud/storage/client_sign_policy_document_test.cc
+++ b/google/cloud/storage/client_sign_policy_document_test.cc
@@ -190,7 +190,6 @@ TEST_F(CreateSignedPolicyDocRPCTest, SignPolicyPermanentFailure) {
       "SignBlob");
 }
 
-#if GOOGLE_CLOUD_CPP_HAVE_CODECVT
 PolicyDocumentV4 CreatePolicyDocumentV4ForTest() {
   PolicyDocumentV4 result;
   result.bucket = "test-bucket";
@@ -296,7 +295,6 @@ TEST_F(CreateSignedPolicyDocTest, SignV4VirtualHostname) {
 
   EXPECT_EQ("https://test-bucket.storage.googleapis.com/", actual->url);
 }
-#endif  // GOOGLE_CLOUD_CPP_HAVE_CODECVT
 
 }  // namespace
 }  // namespace STORAGE_CLIENT_NS

--- a/google/cloud/storage/internal/policy_document_request.cc
+++ b/google/cloud/storage/internal/policy_document_request.cc
@@ -17,17 +17,10 @@
 #include "google/cloud/storage/internal/openssl_util.h"
 #include "google/cloud/internal/format_time_point.h"
 #include <nlohmann/json.hpp>
-#if GOOGLE_CLOUD_CPP_HAVE_CODECVT
 #include <codecvt>
-#include <locale>
-#endif  // GOOGLE_CLOUD_CPP_HAVE_CODECVT
 #include <iomanip>
+#include <locale>
 #include <sstream>
-
-// Some MSVC versions need this.
-#if (!_DLL) && (_MSC_VER >= 1900) && (_MSC_VER <= 1911)
-std::locale::id std::codecvt<char16_t, char, _Mbstatet>::id;
-#endif
 
 namespace google {
 namespace cloud {
@@ -97,12 +90,21 @@ bool EscapeAsciiChar(std::string& result, char32_t c) {
 }
 }  // namespace
 
-#if GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 StatusOr<std::string> PostPolicyV4EscapeUTF8(std::string const& utf8_bytes) {
   std::string result;
 
-  std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> conv;
-  std::basic_string<char32_t> utf32;
+#if (_MSC_VER >= 1900)
+  // Working around missing std::codecvt_utf8<char32_t> symbols in MSVC
+  // Microsoft bug number: VSO#143857
+  // Context:
+  // https://social.msdn.microsoft.com/Forums/en-US/8f40dcd8-c67f-4eba-9134-a19b9178e481/vs-2015-rc-linker-stdcodecvt-error?forum=vcgeneral
+  using WideChar = __int32;
+#else   // (_MSC_VER >= 1900)
+  using WideChar = char32_t;
+#endif  // (_MSC_VER >= 1900)
+  std::wstring_convert<std::codecvt_utf8<WideChar>, WideChar> conv;
+  std::basic_string<WideChar> utf32;
   try {
     utf32 = conv.from_bytes(utf8_bytes);
   } catch (std::exception const& ex) {
@@ -110,7 +112,7 @@ StatusOr<std::string> PostPolicyV4EscapeUTF8(std::string const& utf8_bytes) {
                   std::string("string failed to parse as UTF-8: ") + ex.what());
   }
   utf32 = conv.from_bytes(utf8_bytes);
-  for (char32_t c : utf32) {
+  for (auto c : utf32) {
     bool is_ascii = EscapeAsciiChar(result, c);
     if (!is_ascii) {
       // All unicode characters should be encoded as \udead.
@@ -122,13 +124,13 @@ StatusOr<std::string> PostPolicyV4EscapeUTF8(std::string const& utf8_bytes) {
   }
   return result;
 }
-#else   // GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#else   // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 StatusOr<std::string> PostPolicyV4EscapeUTF8(std::string const&) {
   return Status(StatusCode::kUnimplemented,
                 "Signing POST policies is unavailable with this compiler due "
-                "to the lack of `codecvt` header or exception support.");
+                "to the lack of exception support.");
 }
-#endif  // GOOGLE_CLOUD_CPP_HAVE_CODECVT
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 
 StatusOr<std::string> PostPolicyV4Escape(std::string const& utf8_bytes) {
   std::string result;

--- a/google/cloud/storage/internal/policy_document_request.h
+++ b/google/cloud/storage/internal/policy_document_request.h
@@ -149,12 +149,6 @@ class PolicyDocumentV4Request {
 
 std::ostream& operator<<(std::ostream& os, PolicyDocumentV4Request const& r);
 
-#if defined(_MSC_VER)
-#define GOOGLE_CLOUD_CPP_HAVE_CODECVT 0
-#else
-#define GOOGLE_CLOUD_CPP_HAVE_CODECVT 1
-#endif
-
 }  // namespace internal
 }  // namespace STORAGE_CLIENT_NS
 }  // namespace storage

--- a/google/cloud/storage/internal/policy_document_request_test.cc
+++ b/google/cloud/storage/internal/policy_document_request_test.cc
@@ -59,7 +59,7 @@ TEST(PostPolicyV4EscapeTest, OnlyAscii) {
             *PostPolicyV4Escape("\\\b\f\n\r\t\vabcd"));
 }
 
-#if GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 TEST(PostPolicyV4EscapeTest, InvalidUtf) {
   // In UTF8 a byte larger than 0x7f indicates that there is a byte following
   // it. Here we truncate the string to cause the UTF parser to fail.
@@ -67,17 +67,17 @@ TEST(PostPolicyV4EscapeTest, InvalidUtf) {
   EXPECT_EQ(StatusCode::kInvalidArgument,
             PostPolicyV4Escape(invalid_utf8).status().code());
 }
-#endif  // GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 
 TEST(PostPolicyV4EscapeTest, Simple) {
-#if GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
   EXPECT_EQ("\127\065abcd$", *PostPolicyV4Escape("\127\065abcd$"));
   EXPECT_EQ("\\\\\\b\\f\\n\\r\\t\\v\\u0080\\u0119",
-            *PostPolicyV4Escape("\\\b\f\n\r\t\v\u0080ę"));
-#else   // GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+            *PostPolicyV4Escape(u8"\\\b\f\n\r\t\v\u0080\u0119"));
+#else   // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
   EXPECT_EQ(StatusCode::kUnimplemented,
             PostPolicyV4Escape("ąę").status().code());
-#endif  // GOOGLE_CLOUD_CPP_HAVE_CODECVT && GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
 TEST(PolicyDocumentV4Request, Printing) {

--- a/google/cloud/storage/tests/signed_url_conformance_test.cc
+++ b/google/cloud/storage/tests/signed_url_conformance_test.cc
@@ -442,13 +442,12 @@ int main(int argc, char* argv[]) {  // NOLINT(bugprone-exception-escape)
                  back_inserter(name), [](char c) {
                    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
                  });
-#if !GOOGLE_CLOUD_CPP_HAVE_CODECVT || !GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#if !GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
     if (name == "POSTPolicyCharacterEscaping") {
-      // Escaping is not supported if codecvt header or exceptions are
-      // unavailable.
+      // Escaping is not supported if exceptions are unavailable.
       continue;
     }
-#endif  // !GOOGLE_CLOUD_CPP_HAVE_CODECVT || !GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#endif  // !GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
     bool inserted =
         google::cloud::storage::post_policy_tests->emplace(name, j_obj.value())
             .second;

